http-mclient/src/Network/HTTP/MClient.hs

```diff
@@ -138,16 +140,55 @@ send' transactionDecoder decoder request = do
   sendRequest sendRequest' transactionDecoder decoder request
 
 sendRequest
-  :: SendRequest
+  :: forall env a . Env env
+  => SendRequest
   -> TransactionDecoder a
   -> ResponseDecoder a
   -> HTTP.Request
   -> MIO env (Result a)
 sendRequest sendRequest' transactionDecoder decoder request =
-  either transactionDecoder decoder
-    <$> Exception.tryJust selectException (liftIO $ sendRequest' request)
+  either transactionDecoder decoder <$> retriableSendRequest
   where
-    selectException :: HTTP.HttpException -> Maybe HTTP.HttpExceptionContent
-    selectException = \case
-      (HTTP.HttpExceptionRequest _request content) -> pure content
-      _other                                       -> empty
+    retriableSendRequest :: MIO env (Either HTTP.HttpExceptionContent (HTTP.Response LBS.ByteString))
+    retriableSendRequest
+      = Retry.retrying
+        limitedBackoff
+        isRetriable
+      $ const (Exception.tryJust selectException . liftIO $ sendRequest' request)
+      where
+        limitedBackoff :: Retry.RetryPolicy
+        limitedBackoff = Retry.exponentialBackoff 50_000 <> Retry.limitRetries (convertImpure retryLimit)
+
+        isRetriable
+          :: Retry.RetryStatus
+          -> Either HTTP.HttpExceptionContent (HTTP.Response LBS.ByteString)
+          -> MIO env Bool
+        isRetriable Retry.RetryStatus{..} = either process (const (pure False))
+          where
+            process :: HTTP.HttpExceptionContent -> MIO env Bool
+            process error = if isRequestNotSent error then logError $> True else pure False
+              where
+                isRequestNotSent = \case
+                  HTTP.ConnectionTimeout            -> True
+                  HTTP.ConnectionFailure _exception -> True
+                  _                                 -> False
```

Comment:
__comment:__ Given it's a common question, I would probably add a comment stating both of these kinds of errors are only returned during connection establishment; so we can be certain nothing has been sent yet when we get these.

Reply:
I've independently read the `http-client` source code and have verified that `Connection{Timeout,Failure}` are only emitted from making new connections not yet having send anything to the server.

Now: we set a response timeout intentionally, so re-trying a connection timeout: (which gets taken out of the response timeout budet) is a bad idea. As we use these timeouts intentionally to make sure the system does not stal.

This leaves only `ConnectionFailure` as the only low level re-tryable error.

Reply:
> Now: we set a response timeout intentionally, so re-trying a connection timeout: (which gets taken out of the response timeout budet) is a bad idea. As we use these timeouts intentionally to make sure the system does not stal.

This rationale makes sense. I will remove it.